src/fs/rock/RockRebuild.cc

```diff
@@ -694,19 +697,41 @@ Rock::Rebuild::swanSong()
     storeRebuildComplete(&counts);
 }
 
+void
+Rock::Rebuild::callException(const std::exception &e)
+{
+    ErrorDetail::Pointer errorDetail;
+    if (const auto tex = dynamic_cast<const TextException*>(&e))
+        errorDetail = new ExceptionErrorDetail(tex->id());
+    else
+        errorDetail = new ExceptionErrorDetail(Here().id());
+    failure(ToSBuf(e.what(),
+        Debug::Extra, errorDetail).c_str());
+    // fatal failure() substitutes default AsyncJob::callException(e) handling
+}
+
 void
 Rock::Rebuild::failure(const char *msg, int errNo)
 {
+    assert(sd);
     debugs(47,5, sd->index << " slot " << loadingPos << " at " <<
            dbOffset << " <= " << dbSize);
 
-    if (errNo)
-        debugs(47, DBG_CRITICAL, "ERROR: Rock cache_dir rebuild failure: " << xstrerr(errNo));
-    debugs(47, DBG_CRITICAL, "Do you need to run 'squid -z' to initialize storage?");
+    auto error = ToSBuf("Cannot rebuild rock cache_dir index for ", sd->filePath,
+        Debug::Extra, "problem: ", msg,
+        Debug::Extra, "scan progress: ", Math::int64Percent(loadingPos, dbSlotLimit), '%');
+
+    if (errNo) {
+        auto detail = SysErrorDetail::NewIfAny(errNo);
+        error.append(ToSBuf(Debug::Extra, detail));
+/*        error.append(ToSBuf(Debug::Extra, "I/O error: ", xstrerr(errNo)));
+        // XXX: squid -z is only useful for ENOENT, but ENOENT is detected in
```

Comment:
If this XXX is accurate, then we should remove the "squid -z" hint (already done) and explain why it was removed (i.e., retell this XXX) in the PR description (not done).

Reply:
I adjusted PR description.